src/ecs/resources/input/button.rs

```diff
@@ -0,0 +1,47 @@
+use engine::{VirtualKeyCode, MouseButton};
+
+use super::local_mouse_button::LocalMouseButton;
+use super::local_virtual_key_code::LocalVirtualKeyCode;
+
+/// A Button is any kind of digital input that the engine supports.
+#[derive(Eq, PartialEq, Debug, Copy, Clone, Serialize, Deserialize)]
+pub enum Button {
+    /// Keyboard keys
+    Key(#[serde(with = "LocalVirtualKeyCode")]
+        VirtualKeyCode),
+
+    /// Mouse buttons
+    Mouse(#[serde(with = "LocalMouseButton")]
+          MouseButton),
+    //TODO: Add controller buttons here when the engine has support.
+}
+
+impl From<VirtualKeyCode> for Button {
+    fn from(keycode: VirtualKeyCode) -> Self {
+        Button::Key(keycode)
+    }
+}
+
+impl From<MouseButton> for Button {
+    fn from(mouse_button: MouseButton) -> Self {
+        Button::Mouse(mouse_button)
+    }
+}
+
+/// Describes an input state for a button.
+#[derive(Eq, PartialEq, Debug, Copy, Clone)]
+pub enum ButtonState {
```

Comment:
I like the structure here :+1: We should probably think about at which level we want to abstract over different input methods and we should design a clear and obvious API for that.